File: javascript/merge_sort.js
function mergeSort(list) {
    // If the list contains 0 or 1 , it is already sorted
    if (list.length <= 1) {
        return list
    }

    // Split the list into two halves
    const {firstHalf, secondHalf } = splitArray(list)

    // Recursively sort both halves:
    const left = mergeSort(firstHalf)
    const right = mergeSort(secondHalf)

    // Merge the sorted halves back together:
    return merge(left, right)

}

// Find the midpoint to split the array into two halves and returns them
function splitArray(array) {
    const midpoint = Math.floor(array.length / 2)

    const firstHalf = array.slice(0, midpoint)
    const secondHalf = array.slice(midpoint)

    return {firstHalf, secondHalf}

}

function merge(left, right) {
    let l = []
    // Initialize pointers for both left and right arrays.
    let i = 0
    let j = 0

    while (i < left.length && j < right.length) {
        // If the current element in the left array is smaller, add it to the result.
        if (left[i] < right [j]) {
            l.push(left[i])
            i += 1
        // Otherwise, add the current element in the right array to the result.
        } else {
            l.push(right[j])
            j += 1
        }
    }

    // Add any remaining elements from the left array to the result.
    while (i < left.length) {
        l.push(left[i])
        i += 1
    }
    
    // Add any remaining elements from the right array to the result.
    while (j < right.length) {
        l.push(right[j])
        j += 1
    }
    return l;
}

const alist = [54, 26, 62, 93, 77, 32, 44, 55, 20]
const l = mergeSort(alist)

console.log(l)

File: python/merge_sort.py
def merge_sort(list):
    """
    Sorts a list in ascending order
    Returns a new sorted list

    Divide: Find the midpoint of the list and divide into sublists
    Conquer: Recursively sort the sublists created in previous steps
    Combine: Merge the sorted sublists created in previous step

    The run time of the overall process is O(k n log n)
    Take linear space of 0(n)
    """

    if len(list) <= 1:
        return list
    
    left_half, right_half = split(list)
    left = merge_sort(left_half)
    right = merge_sort(right_half)
   
    return merge(left, right)


def split(list):
    """
    Divide the unsorted list at mipoint into sublists
    Return two sublists - left and right

    Takes overall O(k log n) time
    """

    mid = len(list)//2
    left = list[:mid]  # split actually takes O(k), can use recursive binary
    # search to solve this
    right = list[mid:]  # split actually takes O(k), can use recursive binary
    # search to solve this

    return left, right


def merge(left, right):
    """
    Merges two lists (arrays), sorting them in the process
    Return a new merged list

    Takes overall O(n) time
    """

    li = []
    i = 0
    j = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]: 
            li.append(left[i])
            i += 1
        else:
            li.append(right[j])
            j += 1

    while i < len(left):
        li.append(left[i])
        i += 1

    while j < len(right):
        li.append(right[j])
        j += 1

    return li


# print(l)

def verify_sorted(list):
    n = len(list)

    if n == 0 or n == 1:
        return True

    return list[0] < list[1] and verify_sorted(list[1:])


alist = [54, 26, 62, 93, 77, 32, 44, 55, 20]
li = merge_sort(alist)
print(verify_sorted(alist))
print(verify_sorted(li))
